doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1160,6 +1173,388 @@ Printing |Ltac| tactics
 
    This command displays a list of all user-defined tactics, with their arguments.
 
+
+.. _ltac-examples:
+
+Examples of using |Ltac|
+-------------------------
+
+
+About the cardinality of the set of natural numbers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The first example which shows how to use pattern matching over the
+proof context is a proof of the fact that natural numbers have more
+than two elements. This can be done as follows:
+
+.. coqtop:: in reset
+
+   Lemma card_nat :
+     ~ exists x : nat, exists y : nat, forall z:nat, x = z \/ y = z.
+   Proof.
+
+.. coqtop:: in
+
+   red; intros (x, (y, Hy)).
+
+.. coqtop:: in
+
+   elim (Hy 0); elim (Hy 1); elim (Hy 2); intros;
+
+   match goal with
+       | _ : ?a = ?b, _ : ?a = ?c |- _ =>
+           cut (b = c); [ discriminate | transitivity a; auto ]
+   end.
+
+.. coqtop:: in
+
+   Qed.
+
+We can notice that all the (very similar) cases coming from the three
+eliminations (with three distinct natural numbers) are successfully
+solved by a match goal structure and, in particular, with only one
+pattern (use of non-linear matching).
+
+
+Permutations of lists
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A more complex example is the problem of permutations of
+lists. The aim is to show that a list is a permutation of
+another list.
+
+.. coqtop:: in reset
+
+   Section Sort.
+
+.. coqtop:: in
+
+   Variable A : Set.
+
+.. coqtop:: in
+
+   Inductive perm : list A -> list A -> Prop :=
+       | perm_refl : forall l, perm l l
+       | perm_cons : forall a l0 l1, perm l0 l1 -> perm (a :: l0) (a :: l1)
+       | perm_append : forall a l, perm (a :: l) (l ++ a :: nil)
+       | perm_trans : forall l0 l1 l2, perm l0 l1 -> perm l1 l2 -> perm l0 l2.
+
+.. coqtop:: in
+
+   End Sort.
+
+First, we define the permutation predicate as shown above.
+
+.. coqtop:: none
+
+   Require Import List.
+
+
+.. coqtop:: in
+
+   Ltac perm_aux n :=
+   match goal with
+       | |- (perm _ ?l ?l) => apply perm_refl
+       | |- (perm _ (?a :: ?l1) (?a :: ?l2)) =>
+           let newn := eval compute in (length l1) in
+               (apply perm_cons; perm_aux newn)
+       | |- (perm ?A (?a :: ?l1) ?l2) =>
+           match eval compute in n with
+               | 1 => fail
+               | _ =>
+                   let l1' := constr:(l1 ++ a :: nil) in
+                       (apply (perm_trans A (a :: l1) l1' l2);
+                       [ apply perm_append | compute; perm_aux (pred n) ])
+           end
+   end.
+
+Next we define an auxiliary tactic ``perm_aux`` which takes an argument
+used to control the recursion depth. This tactic behaves as follows. If
+the lists are identical (i.e. convertible), it concludes. Otherwise, if
+the lists have identical heads, it proceeds to look at their tails.
+Finally, if the lists have different heads, it rotates the first list by
```

Comment:
```suggestion
If the lists have different heads, it rotates the first list by
```

It might be more effective to format this as pseudocode that mimics the structure of the Ltac.  Maybe add comments for much of this to the Ltac script; it will be easier to read rather than having 2 chunks of material for the user to correlate.